### PR TITLE
Segformer model and recipe

### PR DESCRIPTION
**Pull request by eran-deci**

### CODE REVIEW

File: src/super_gradients/training/models/segmentation_models/segformer.py [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+        # Transform to proper mean, std
+        tensor.mul_(std * math.sqrt(2.0))
+        tensor.add_(mean)
+
+        # Clamp to ensure it's in the proper range
+        tensor.clamp_(min=a, max=b)
+        return tensor
+
+
+def trunc_normal_(tensor, mean=0.0, std=1.0, a=-2.0, b=2.0):
```
Comment (BloodAxe) [inline src/super_gradients/training/models/segmentation_models/segformer.py:60]: Please consider using https://pytorch.org/docs/stable/nn.init.html#torch.nn.init.trunc_normal_ instead as it's probably same implementation, but library method is preferrable.

File: src/super_gradients/training/models/segmentation_models/segformer.py [hunk 2]
```diff
@@ -0,0 +1,599 @@
...
+
+        attn = (q @ k.transpose(-2, -1)) * self.scale
+        attn = attn.softmax(dim=-1)
+
+        x = (attn @ v).transpose(1, 2).reshape(b, n, c)
+        x = self.proj(x)
+        return x
+
+
+class DropPath(nn.Module):
```
Comment (BloodAxe) [inline src/super_gradients/training/models/segmentation_models/segformer.py:148]: Possible duplicate of super_gradients.training.utils.regularization_utils.DropPath

File: src/super_gradients/training/models/segmentation_models/segformer.py [hunk 3]
```diff
@@ -0,0 +1,599 @@
...
+        :param eff_self_att_heads: number of efficient self-attention heads in each stage
+        :param overlap_patch_size:  the patch size of the overlapping patch embedding in each stage
+        :param overlap_patch_stride:  the patch stride of the overlapping patch embedding in each stage
+        :param overlap_patch_pad:  the patch padding of the overlapping patch embedding in each stage
+        :param in_channels:  number of input channels
+        """
+
+        super().__init__()
+
+        assert (
```
Comment (BloodAxe) [inline src/super_gradients/training/models/segmentation_models/segformer.py:254]: Rule of thumb: assert in unit tests, if/raise in the code. 
WHY: https://github.com/Deci-AI/super-gradients/pull/599#discussion_r1116999292

File: src/super_gradients/recipes/dataset_params/cityscapes_segformer_dataset_params.yaml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        crop_size: [ 1024, 1024 ]
+        fill_mask: 19
+
+    - SegCropImageAndMask:
+        crop_size: [ 1024, 1024 ]
+        mode: random
+
+val_dataset_params:
+  transforms:
+    - SegRescale:
```
Comment (BloodAxe) [inline src/super_gradients/recipes/dataset_params/cityscapes_segformer_dataset_params.yaml:28]: If I read this transformation chain correctly, then we
1) Resize input image to ensure smallest size is 1024
2) Pad image to 1024x1024 (Looks it it redundant because we just did resize than guarantee that smallest size is no smaller than 1024)
3) Center crop. This step looks wrong to me as we validate only on central part of the image.

Perhaps, instead we want to do
1) Resize input image to ensure **longest** size is 1024
2) Pad to 1024x1024
 
Please let me know what you think or did I miss anything

File: src/super_gradients/training/models/segmentation_models/segformer.py (hunk 2 above)
Comment (eran-deci) [reply src/super_gradients/training/models/segmentation_models/segformer.py]: Changed to SG implementation

File: src/super_gradients/training/models/segmentation_models/segformer.py (hunk 3 above)
Comment (eran-deci) [reply src/super_gradients/training/models/segmentation_models/segformer.py]: Changed to if/raise

**Comment by BloodAxe** [comment]
Hi @eran-deci do you mind updating the PR to make the build passing? Looks like there is only black formatting that is left

File: src/super_gradients/training/models/segmentation_models/segformer.py [hunk 5]
```diff
@@ -0,0 +1,516 @@
...
+        self.kv = nn.Linear(dim, dim * 2)
+        self.proj = nn.Linear(dim, dim)
+
+        if sr_ratio > 1:
+            self.sr = nn.Conv2d(dim, dim, sr_ratio, sr_ratio)
+            self.norm = nn.LayerNorm(dim)
+
+    def forward(self, x: torch.Tensor, h: int, w: int) -> torch.Tensor:
+        b, n, c = x.shape
+        q = self.q(x).reshape(b, n, self.head, c // self.head).permute(0, 2, 1, 3)
```
Comment (ofrimasad) [inline src/super_gradients/training/models/segmentation_models/segformer.py:72]: Did this code successfully convert to ONNX?

File: src/super_gradients/training/models/segmentation_models/segformer.py [hunk 6]
```diff
@@ -0,0 +1,516 @@
...
+
+    def forward(self, x: torch.Tensor, h: int, w: int) -> torch.Tensor:
+        x = self.fc1(x)
+
+        b, _, c = x.shape
+        x = x.transpose(1, 2).view(b, c, h, w)
+        x = self.dwconv(x)
+        x = x.flatten(2).transpose(1, 2)
+
+        x = self.fc2(F.gelu(x))
```
Comment (ofrimasad) [inline src/super_gradients/training/models/segmentation_models/segformer.py:111]: can this code be converted to ONNX? (the GeLU)

File: src/super_gradients/training/models/segmentation_models/segformer.py [hunk 7]
```diff
@@ -0,0 +1,516 @@
...
+            x = x.reshape(b_size, h, w, -1).permute(0, 3, 1, 2)
+
+            features.append(x)
+
+        return features
+
+
+class MLP(nn.Module):
+    def __init__(self, dim, embed_dim):
+        """
```
Comment (ofrimasad) [inline src/super_gradients/training/models/segmentation_models/segformer.py:242]: add a //TODO to move this as well

File: src/super_gradients/training/models/segmentation_models/segformer.py [hunk 8]
```diff
@@ -0,0 +1,516 @@
...
+        self.proj = nn.Linear(dim, embed_dim)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = x.flatten(2).transpose(1, 2)
+        x = self.proj(x)
+
+        return x
+
+
+class LinearFuse(nn.Module):
```
Comment (ofrimasad) [inline src/super_gradients/training/models/segmentation_models/segformer.py:259]: please use the `super_gradients.modules.conv_bn_relu_block.ConvBNReLU` block instead of this

File: src/super_gradients/training/models/segmentation_models/segformer.py [hunk 9]
```diff
@@ -0,0 +1,516 @@
...
+
+        return out
+
+
+# TODO: add support for aux heads? (not in original impl) (currently not using)
+class SegFormer(SegmentationModule):
+    def __init__(
+        self,
+        num_classes: int,
+        encoder_embed_dims: list,
```
Comment (ofrimasad) [inline src/super_gradients/training/models/segmentation_models/segformer.py:318]: please use full typing on all of the params:
`List[int]` or `List[float]` instead of `list`

File: src/super_gradients/training/models/segmentation_models/segformer.py [hunk 10]
```diff
@@ -0,0 +1,516 @@
...
+                    nn.init.constant_(m.bias, 0)
+            elif isinstance(m, (nn.LayerNorm, nn.BatchNorm2d, nn.SyncBatchNorm)):
+                nn.init.ones_(m.weight)
+                nn.init.zeros_(m.bias)
+
+    @property
+    def backbone(self):
+        return self._backbone
+
+    def _remove_auxiliary_heads(self):
```
Comment (ofrimasad) [inline src/super_gradients/training/models/segmentation_models/segformer.py:378]: why do we need this?

**Review by ofrimasad — COMMENTED** [review]
Very nice work.
Was this code taken from the original repo or another repo?
please make sure the code has Apache/BSD/MIT license 
My comments are inline. we will fix some more general SG related comment in the future (just add the # TODO in the required places)

File: src/super_gradients/training/models/all_architectures.py [hunk 11]
```diff
@@ -144,7 +152,12 @@
     Models.PP_YOLOE_M: PPYoloE_M,
     Models.PP_YOLOE_L: PPYoloE_L,
     Models.PP_YOLOE_X: PPYoloE_X,
-    #
+    Models.SEGFORMER_B0: SegFormerB0,
```
[the remark this answers is not in the document]
Comment (eran-deci) [reply src/super_gradients/training/models/all_architectures.py:155]: any action item on my end here? (seems in master this is still the format)

File: src/super_gradients/training/models/segmentation_models/segformer.py (hunk 10 above)
Comment (eran-deci) [reply src/super_gradients/training/models/segmentation_models/segformer.py:378]: it is an unimplemented abstract method, and it's simply 'pass' since currently aux heads aren't implemented for segformer

File: src/super_gradients/recipes/cityscapes_segformer_b0.yaml [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+#   1. We recommend preparing the data according to SG's CityScapes readme file:
+#      https://github.com/Deci-AI/super-gradients/blob/master/src/super_gradients/training/datasets/Dataset_Setup_Instructions.md
+#   2. Note: if you change the dataset's internal directory structure, make changes to the fields "list_file" and
+#      "labels_csv_path" of both "train_dataset_params" and "val_dataset_params" accordingly
+#   3. Edit the "data_root_dir" field below to point to the absolute path of the data root directory
+#   4. Edit the "ckpt_root_dir" field to the path where you want to save checkpoints and logs
+#   5. Move to the project root (where you will find the ReadMe and src folder)
+#   6. Run the command:
+#       python src/super_gradients/examples/train_from_recipe_example/train_from_recipe.py --config-name=cityscapes_segformer_b0
+#
```
[the remark this answers is not in the document]
Comment (eran-deci) [reply src/super_gradients/recipes/cityscapes_segformer_b0.yaml:23]: done

File: src/super_gradients/recipes/cityscapes_segformer_b1.yaml [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+#
+# Imagenet-1K pre-trained backbone:
+#   MiT (Mix Transformer) B1:   https://deci-pretrained-models.s3.amazonaws.com/mit_backbones/mit_b1.pth
+#
+#   1. Download the weights from the above link and put them in a directory of your choice
+#   2. Below, insert the weights file's full path to checkpoint_params.checkpoint_path
+#   3. Ensure checkpoint_params.load_backbone: True
+
+
+defaults:
```
[the remark this answers is not in the document]
Comment (eran-deci) [reply src/super_gradients/recipes/cityscapes_segformer_b1.yaml:35]: done

File: src/super_gradients/training/models/segmentation_models/segformer.py [hunk 14]
```diff
@@ -0,0 +1,516 @@
...
+            else:
+                multiply_lr_params[name] = param
+        return multiply_lr_params.items(), no_multiply_params.items()
+
+
+class SegFormerCustom(SegFormer):
+    def __init__(self, arch_params: HpmStruct):
+        """Parse arch_params and translate the parameters to build the SegFormer architecture"""
+        super().__init__(
+            num_classes=arch_params.num_classes,
```
[the remark this answers is not in the document]
Comment (eran-deci) [reply src/super_gradients/training/models/segmentation_models/segformer.py:430]: done

File: src/super_gradients/training/models/segmentation_models/segformer.py [hunk 15]
```diff
@@ -0,0 +1,516 @@
...
+        _, _, h, w = x.shape
+
+        x = x.flatten(2).transpose(1, 2)
+        x = self.norm(x)
+
+        return x, h, w
+
+
+class EfficientSelfAttention(nn.Module):
+    def __init__(self, dim: int, head: int, sr_ratio: int):
```
[the remark this answers is not in the document]
Comment (eran-deci) [reply src/super_gradients/training/models/segmentation_models/segformer.py:49]: done

File: src/super_gradients/training/models/segmentation_models/segformer.py (hunk 7 above)
Comment (eran-deci) [reply src/super_gradients/training/models/segmentation_models/segformer.py:242]: done

File: src/super_gradients/training/models/segmentation_models/segformer.py (hunk 8 above)
Comment (eran-deci) [reply src/super_gradients/training/models/segmentation_models/segformer.py:259]: done

File: src/super_gradients/training/models/segmentation_models/segformer.py (hunk 5 above)
Comment (eran-deci) [reply src/super_gradients/training/models/segmentation_models/segformer.py:77]: The whole model was succesfully converted to ONNX, but I didn't check the performance of the ONNX model

File: src/super_gradients/training/models/segmentation_models/segformer.py (hunk 6 above)
Comment (eran-deci) [reply src/super_gradients/training/models/segmentation_models/segformer.py:116]: see above comment

File: src/super_gradients/training/models/segmentation_models/segformer.py (hunk 10 above)
Comment (eran-deci) [reply src/super_gradients/training/models/segmentation_models/segformer.py:367]: @ofrimasad also, please have a look at the failed test, not sure what's the cause

File: src/super_gradients/training/models/all_architectures.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (eran-deci) [reply src/super_gradients/training/models/all_architectures.py]: done

**Review by ofrimasad — APPROVED** [review]
LGTM
